k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapConnection.java

```diff
@@ -356,6 +371,46 @@ private void authenticate() throws MessagingException, IOException {
         }
     }
 
+    private void authXoauth2withSASLIR() throws IOException, MessagingException {
+        try {
+            attemptXOAuth2();
+        } catch (NegativeImapResponseException e) {
+            //We could avoid this if we had a reasonable chance of knowing
+            //if a token was invalid before use (e.g. due to expiry). But we don't
+            //This is the intended behaviour per AccountManager
+            Log.v(LOG_TAG, "Authentication exception, invalidating token and re-trying", e);
+            oauthTokenProvider.invalidateToken(settings.getUsername());
+            try {
+                attemptXOAuth2();
+            } catch (NegativeImapResponseException e2) {
+                //Okay, we failed on a new token.
+                //Invalidate the token anyway but assume it's permanent.
+                Log.v(LOG_TAG, "Authentication exception for new token, permanent error assumed", e);
+                oauthTokenProvider.invalidateToken(settings.getUsername());
+                throw new AuthenticationFailedException(e2.getMessage(), e);
+            }
+        }
+    }
+
+    private void attemptXOAuth2() throws MessagingException, IOException {
+        String command = Commands.AUTHENTICATE_XOAUTH2;
```

Comment:
putting this in a variable seems unnecessary, maybe put the `getToken` output in a variable instead to make the sendSaslIrCommand call more clear?